### PR TITLE
SDK changes for fetch task and submit action

### PR DESCRIPTION
**Pull request by jinxiwu**

### CODE REVIEW

File: Node/src/botbuilder-teams.d.ts [hunk 1]
```diff
@@ -1036,6 +1038,18 @@ export class TeamsChatConnector extends builder.ChatConnector {
   *  @param handler The function to execute when an invoke request to submit task module results is received.
   */
   public onTaskModuleSubmit(handler: TaskModuleSubmitHandlerType): void;
+
+  /**
+  *  Set a handler that is called when an invoke request to fetch task module from compose extension create command is received .
+  *  @param handler The function to execute when an invoke request to fetch task module from compose extension create command is received.
```
Comment (robin-liao) [inline Node/src/botbuilder-teams.d.ts]: "command actions". Avoid using "create"

**Comment by robin-liao** [comment]
Please test it locally using `npm link` to ensure e2e working.

File: Node/src/botbuilder-teams.d.ts [hunk 2]
```diff
@@ -887,6 +887,8 @@ export type SigninStateVerificationHandlerType = (event: builder.IEvent, query:
 export type FileConsentCardResponseHandlerType = (event: builder.IEvent, response: IFileConsentCardResponse, callback: (err: Error, result: any, statusCode?: number) => void) => void;
 export type TaskModuleFetchHandlerType = (event: builder.IEvent, request: ITaskModuleInvokeRequest, callback: (err: Error, result: ITaskModuleResponseOfFetch, statusCode?: number) => void) => void;
 export type TaskModuleSubmitHandlerType = (event: builder.IEvent, request: ITaskModuleInvokeRequest, callback: (err: Error, result: ITaskModuleResponseOfSubmit, statusCode?: number) => void) => void;
+export type ComposeExtensionFetchForActionCommandHandlerType = (event: builder.IEvent, request: ComposeExtensionQuery, callback: (err: Error, result: ITaskModuleResponseOfFetch, statusCode?: number) => void) => void;
```
Comment (maoyu-dev) [inline Node/src/botbuilder-teams.d.ts]: for both fetch and submit, the response could be task module, or compose extension, so you would need a union type of both.

File: Node/src/botbuilder-teams.d.ts [hunk 3]
```diff
@@ -1036,6 +1038,18 @@ export class TeamsChatConnector extends builder.ChatConnector {
   *  @param handler The function to execute when an invoke request to submit task module results is received.
   */
   public onTaskModuleSubmit(handler: TaskModuleSubmitHandlerType): void;
+
+  /**
+  *  Set a handler that is called when an invoke request to fetch task module from compose extension create command is received .
+  *  @param handler The function to execute when an invoke request to fetch task module from compose extension create command is received.
+  */
+ public onComposeExtensionFetch(handler: ComposeExtensionFetchForActionCommandHandlerType): void;
```
Comment (maoyu-dev) [inline Node/src/botbuilder-teams.d.ts]: onComposeExtensionFetchTask

File: Node/src/botbuilder-teams.d.ts [hunk 4]
```diff
@@ -1036,6 +1038,18 @@ export class TeamsChatConnector extends builder.ChatConnector {
...
+
+ /**
+ *  Set a handler that is called when an invoke request to submit action is received .
+ *  @param handler The function to execute when an invoke request to submit action results is received.
+ */
+ public onComposeExtensionSubmit(handler: ComposeExtensionSubmitForActionCommandHandlerType): void;
```
Comment (maoyu-dev) [inline Node/src/botbuilder-teams.d.ts]: onComposeExtensionSubmitAction

Comment (jinxiwu) [reply Node/src/botbuilder-teams.d.ts]: fixed

File: Node/src/TeamsChatConnector.ts [hunk 5]
```diff
@@ -48,6 +48,9 @@ export type SigninStateVerificationHandlerType = (event: builder.IEvent, query:
 export type FileConsentCardResponseHandlerType = (event: builder.IEvent, response: IFileConsentCardResponse, callback: (err: Error, result: any, statusCode?: number) => void) => void;
 export type TaskModuleFetchHandlerType = (event: builder.IEvent, request: task.ITaskModuleInvokeRequest, callback: (err: Error, result: task.ITaskModuleResponseOfFetch, statusCode?: number) => void) => void;
 export type TaskModuleSubmitHandlerType = (event: builder.IEvent, request: task.ITaskModuleInvokeRequest, callback: (err: Error, result: task.ITaskModuleResponseOfSubmit, statusCode?: number) => void) => void;
+export type ComposeExtensionFetchForActionCommandHandlerType = (event: builder.IEvent, request: IComposeExtensionActionCommandRequest, callback: (err: Error, result: task.ITaskModuleResponseOfFetch | IComposeExtensionResponse, statusCode?: number) => void) => void;
+export type ComposeExtensionSubmitForActionCommandHandlerType = (event: builder.IEvent, request: IComposeExtensionActionCommandRequest, callback: (err: Error, result: task.ITaskModuleResponseOfFetch | IComposeExtensionResponse, statusCode?: number) => void) => void;
```
Comment (robin-liao) [inline Node/src/TeamsChatConnector.ts]: for submit, result type should be `task.ITaskModuleResponseOfSubmit | IComposeExtensionResponse`

File: Node/src/TeamsChatConnector.ts [hunk 6]
```diff
@@ -358,6 +366,14 @@ export class TeamsChatConnector extends builder.ChatConnector {
     this.taskModuleSubmitHandler = handler;
   }
 
+  public onComposeExtensionFetchTask(handler: ComposeExtensionFetchForActionCommandHandlerType): void {
```
Comment (robin-liao) [inline Node/src/TeamsChatConnector.ts]: make naming consistent across entire PR.

File: Node/src/TeamsChatConnector.ts [hunk 7]
```diff
@@ -358,6 +366,14 @@ export class TeamsChatConnector extends builder.ChatConnector {
...
+    this.composeExtensionFetchForActionCommandHandler = handler;
+  }
+
+  public onComposeExtensionSubmitAction(handler: ComposeExtensionSubmitForActionCommandHandlerType): void {
```
Comment (robin-liao) [inline Node/src/TeamsChatConnector.ts]: make naming consistent across entire PR.

File: Node/src/TeamsChatConnector.ts (hunk 5 above)
Comment (jinxiwu) [reply Node/src/TeamsChatConnector.ts]: done

File: Node/src/TeamsChatConnector.ts (hunk 6 above)
Comment (jinxiwu) [reply Node/src/TeamsChatConnector.ts]: done

File: Node/src/TeamsChatConnector.ts (hunk 7 above)
Comment (jinxiwu) [reply Node/src/TeamsChatConnector.ts]: done

File: Node/test/TeamsChatConnector.js [hunk 8]
```diff
@@ -444,4 +444,97 @@ describe('TeamsChatConnector', function () {
       })
     });
   });
-});
\ No newline at end of file
+
+  describe('#onComposeExtensionFetch()', function () {
```
Comment (robin-liao) [inline Node/test/TeamsChatConnector.js]: `#onComposeExtensionFetch()` to `#onComposeExtensionFetchTask()` for naming consistent

File: Node/test/TeamsChatConnector.js [hunk 9]
```diff
@@ -444,4 +444,97 @@ describe('TeamsChatConnector', function () {
...
+        }
+      ], (err, body, status) => {
+        assert.ok(!err, 'An error occurred: ' + err);
+        assert.ok(wasHandlerCalled, 'The registered onInvoke handler was not called');
+        done();
+      })
+    });    
+  });
+  
+  describe('#onComposeExtensionSubmit()', function () {
```
Comment (robin-liao) [inline Node/test/TeamsChatConnector.js]: `#onComposeExtensionSubmit()` to `#onComposeExtensionSubmitAction()` for naming consistent

File: Node/test/TeamsChatConnector.js (hunk 8 above)
Comment (jinxiwu) [reply Node/test/TeamsChatConnector.js]: fixed

File: Node/test/TeamsChatConnector.js (hunk 9 above)
Comment (jinxiwu) [reply Node/test/TeamsChatConnector.js]: fixed

File: Node/test/TeamsChatConnector.js [hunk 10]
```diff
@@ -444,4 +444,97 @@ describe('TeamsChatConnector', function () {
...
+  
+      connector.onDispatchEvents([
+        {
+          type: 'invoke',
+          name: lib.TeamsChatConnector.composeExtensionInvokeNameofFetch,
+          value: { }  
+        }
+      ], (err, body, status) => {
+        assert.ok(!err, 'An error occurred: ' + err);
+        assert.ok(wasHandlerCalled, 'The registered onTaskModuleFetch handler was not called');
```
Comment (robin-liao) [inline Node/test/TeamsChatConnector.js]: The registered **onTaskModuleFetch**...  --> **onComposeExtensionFetchTask**

File: Node/test/TeamsChatConnector.js [hunk 11]
```diff
@@ -444,4 +444,97 @@ describe('TeamsChatConnector', function () {
...
+  
+      connector.onDispatchEvents([
+        {
+          type: 'invoke',
+          name: lib.TeamsChatConnector.composeExtensionInvokeNameofFetch,
+          value: { }
+        }
+      ], (err, body, status) => {
+        assert.ok(!err, 'An error occurred: ' + err);
+        assert.ok(wasHandlerCalled, 'The registered onInvoke handler was not called');
```
Comment (robin-liao) [inline Node/test/TeamsChatConnector.js]: The registered **onInvoke** --> ** onComposeExtensionFetch**

File: Node/test/TeamsChatConnector.js [hunk 12]
```diff
@@ -444,4 +444,97 @@ describe('TeamsChatConnector', function () {
...
+  
+      connector.onDispatchEvents([
+        {
+          type: 'invoke',
+          name: lib.TeamsChatConnector.composeExtensionInvokeNameofSubmit,
+          value: { }  
+        }
+      ], (err, body, status) => {
+        assert.ok(!err, 'An error occurred: ' + err);
+        assert.ok(wasHandlerCalled, 'The registered onTaskModuleFetch handler was not called');
```
Comment (robin-liao) [inline Node/test/TeamsChatConnector.js]: The registered **onTaskModuleFetch** --> ** onComposeExtensionSubmitAction**

File: Node/src/botbuilder-teams.d.ts [hunk 13]
```diff
@@ -769,6 +769,12 @@ export interface IComposeExtensionResponse {
   composeExtension?: ComposeExtensionResult;
 }
 
+/** Represents the request to a compose extension action command invoke message */
```
Comment (aosolis) [inline Node/src/botbuilder-teams.d.ts]: I don't quite understand what this means. Should the documentation be similar to what ITaskModuleInvokeRequest has? ITaskModuleInvokeRequest: /** Represents the value of the invoke activity of task module fetch and submit. */

File: Node/src/botbuilder-teams.d.ts [hunk 14]
```diff
@@ -1036,6 +1044,18 @@ export class TeamsChatConnector extends builder.ChatConnector {
   *  @param handler The function to execute when an invoke request to submit task module results is received.
   */
   public onTaskModuleSubmit(handler: TaskModuleSubmitHandlerType): void;
+
+  /**
+  *  Set a handler that is called when an invoke request from command actions to fetch task is received .
```
Comment (aosolis) [inline Node/src/botbuilder-teams.d.ts]: Hard to understand. Can you rephrase it to:
 *  Set a handler for compose extension invoke request made when <describe situation here>

Comment (aosolis) [reply Node/src/botbuilder-teams.d.ts]: Same for onComposeExtensionFetchTask

File: Node/src/botbuilder-teams.d.ts [hunk 15]
```diff
@@ -1036,6 +1044,18 @@ export class TeamsChatConnector extends builder.ChatConnector {
...
+  *  @param handler The function to execute when an invoke request from command actions to fetch task  is received.
+  */
+  public onComposeExtensionFetchTask(handler: ComposeExtensionFetchTaskHandlerType): void;
```
Comment (aosolis) [inline Node/src/botbuilder-teams.d.ts]: Should this be onComposeExtensionFetchTaskInfo?
The notion of "fetching a task" sounds strange to me, especially in the context of a compose extension action.

File: Node/src/botbuilder-teams.d.ts (hunk 13 above)
Comment (jinxiwu) [reply Node/src/botbuilder-teams.d.ts]: make sense, fixed

File: Node/src/botbuilder-teams.d.ts (hunk 14 above)
Comment (jinxiwu) [reply Node/src/botbuilder-teams.d.ts]: fixed

File: Node/src/botbuilder-teams.d.ts (hunk 15 above)
Comment (jinxiwu) [reply Node/src/botbuilder-teams.d.ts]: fetching a task actually means fetching a task module. I want to be consistent with the invoke name "composeExtension/fetchTask"
